### ISSUE TITLE
Crashes when `isShrinkResources` is set to `true`, it reports "The Crashlytics build ID is missing..."

### ISSUE THREAD

**Issue by chachako** [firebase/firebase-android-sdk#5562]
### Environment

  * Android Studio version: **Android Studio Iguana | 2023.2.1 Canary 13**
  * Firebase Component: **Crashlytics**
  * Component version: **2.9.9 (Crashlytics-Gradle)**

### Problem

Unable to run apk using Crashlytics

#### Steps to reproduce:

1. Use Android Gradle Plugin: `8.3.0-alpha13`
2. Set it to the Release variant
3. Enable `isShrinkResources`
4. Just run the app, and the error will occur

I did some investigation and found that the ID is being searched using the hardcoded string here:

https://github.com/firebase/firebase-android-sdk/blob/master/firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CommonUtils.java#L471

However, after decompiling the generated `.apk` file, I noticed that this string resource does not exist in the `.arsc` file. Interestingly, both the `build/generated/crashlytics/res/release/values/com_google_firebase_crashlytics_mappingfileid.xml` path and content are exists. This leads me to suspect that it is being removed when shrinking resources.

It turned out to be true. When I disabled `isShrinkResources`, the issue disappeared.

What confuses me is that I found many other projects that also have `isShrinkResources` set to `true` and use Crashlytics without any issues. The only visible difference seems to be the version of AGP used. Could this potentially be an issue with AGP?

**Comment by rekaszeru** [comment]
I'm facing the same issue with both 8.3.0-alpha13 and 8.3.0-alpha14 (required by AS 2023.2.1 Canary 14). 
Hacking the missing resources into the codebase doesn't work like it did when this issue occurred 3 years ago, neither does tweaking the `upToDateWhen` of Crashlytics tasks.

**Comment by Mygod** [comment]
8.3.0-alpha11 to alpha14 don't work either. Downgrading to 8.3.0-alpha10 worked for me.

**Comment by argzdev** [comment]
Hey folks, thanks for reaching out and adding the tracker ticket to AGP. Let's keep this open while we wait for AGP to fix this on their side.

**Comment by ricowind** [comment]
Are all of you using strict mode for resource shrinking?
https://developer.android.com/build/shrink-code#strict-reference-checks
if you do, you explicitly need keeps for all resources that are only loaded using getIdentifier calls (in this case, this module should have a keep, there are already some keeps in firebase-common/src/main/res/raw/firebase_common_keep.xml)

**Comment by ricowind** [comment]
As a work around, you should be able to do a keep xml files in res/raw that keeps the missing resources.
The reason this is happening now is that we enabled precise shrinking, that will also removed unused entries in the resource table (including strings)

**Comment by ricowind** [comment]
I think we might have a bug for shrinkMode (which we really should not have done like this).
We currently only parse setting it to strictMode (since safe mode is the default), so if there is one rule going in saying strict, we use strict mode. Can one of you share the apk with and without resource shrinking with me, possibly also the .ap_ file going into the compilation, something like: ./app/build/intermediates/processed_res/release/out/resources-release.ap_

**Comment by ricowind** [comment]
Coped from the R8 bug for completeness, in case everyone is not following there

> ok, so the reason for this is that in R classes we replace resource names with dots in them with _ (because the field names can't have dots in them) The whole resource model in AGP is based on this, so the the names of the resources in the model is converted from com.google.firebase.crashlytics.mapping_file_id to com_google_firebase_crashlytics_mapping_file_id
> 
> The actual name in the resource table is still com.google.firebase.crashlytics.mapping_file_id, so the reflective lookup with getIdentifier works with that. You can change your keep use '_' instead of '.', I validated that this works.
> 
> I will see if we can do the string analysis using the _ version for both string constants and keep rules
> 

The reason we are seeing this now is that precise shrinking (which will also shrink the resource table) is now the default (before, we would never remove strings) - the underlying bug has been there since at least 2019 (that we don't support reflective identification of resource names with dots)

**Comment by zoltish** [comment]
Fwiw, the full `keep.xml` file that works for me is the following, placed under `res/raw/`.

```
<resources xmlns:tools="http://schemas.android.com/tools"
    tools:keep="@string/com.google.firebase.crashlytics.*" />
```

Thanks for posting this issue, saved me a ton of headache! 👍🏾

**Comment by mrober** [comment]
This issue has been fixed upstream in AGP 8.3.0-alpha16 see https://issuetracker.google.com/issues/311837855 for the details.
